### PR TITLE
Next better version of object layout

### PR DESCRIPTION
**Pull request by stepancheg**
TL;DR:
* mutability flag and `RefCell` are merged into `ObjectHeader` type
* `ObjectHeader` reimplements `RefCell` using `unsafe`
* this version uses lesser known DST feature of placing DST type
  in the end of another DST type

In this version `Value` is represented as:

```
pub struct Value(ValueInner);

pub(crate) enum ValueInner {
    None(NoneType),
    Bool(bool),
    Int(i64),
    Other(Rc<ValueHolder<dyn TypedValueDyn>>),
}

pub(crate) struct ValueHolder<T: TypedValueDyn + ?Sized> {
    value: ObjectCell<T>,
}

pub(crate) struct ObjectCell<T: ?Sized> {
    // RefCell-like object with
    // * "frozen for iteration" bit
    // * "frozen" state
}

pub(crate) trait TypedValueDyn { ... }

impl<T: TypedValue> TypedValueDyn for T { ... }
```

Note `ValueHolder` is DST.

So the advantages of this approach are:
* data (and header) offsets of an object are computed without virtual calls
* single field now tracks both freezeness state and borrowing
* freeze for iteration is a regular borrow now (with a flag "for iteration")
* frozen objects are now borrowed without changing borrowed flags

These features are partially code cleanup and partially development towards:
* GC
* cross-thread sharing of frozen objects

In particular, since borrowing no longer changes header bits for
frozen objects, these objects could be shared across threads (if
we replaced `Rc` with `Arc`).

The biggest drawback of this commit is the use of `unsafe`.

### CODE REVIEW

**Comment by stepancheg** [comment]
Rebase and better code organization: explicit

```
struct ObjectCell<T: ?Sized> { ... }
```

is introduced which is a `RefCell`-like object which additionally can store "frozen for iteration" flag and "immutable" state.

**Comment by stepancheg** [comment]
Update to a simpler version of the PR. Now Both mutable and immutable objects have a header, and for immutable objects, the header state is always "immutable". This creates a little memory overhead (one word for immutable objects) but simplifies the implementation.

File: starlark/src/values/cell/error.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Cell-related errors
+
+use std::fmt;
+
+/// Error when borrow failed
```
Comment (damienmg) [inline starlark/src/values/cell/error.rs]: nit: Here and there, add ending dot to the comment.

File: starlark/src/values/error.rs [hunk 2]
```diff
@@ -35,7 +35,6 @@ pub const INTERPOLATION_NOT_ENOUGH_PARAMS_ERROR_CODE: &str = "CV10";
 pub const INTERPOLATION_VALUE_IS_NOT_CHAR_ERROR_CODE: &str = "CV12";
 pub const TOO_MANY_RECURSION_LEVEL_ERROR_CODE: &str = "CV13";
 pub const UNSUPPORTED_RECURSIVE_DATA_STRUCTURE_ERROR_CODE: &str = "CV14";
-pub const CANNOT_MUTATE_DURING_ITERATION_ERROR_CODE: &str = "CV15";
```
Comment (damienmg) [inline starlark/src/values/error.rs]: Use some kind of reservation so we don't reallocate the error code in the future, e.g.:
```
const __RESERVED_CV15: &str = "CV15"
```
